Sources/XiEditor/EditView.swift

```diff
@@ -483,9 +492,42 @@ final class EditView: NSView, NSTextInputClient, TextPlaneDelegate {
             } else {
                 let builder = TextLineBuilder(line.text, font: font)
                 builder.setFgColor(argb: foregroundArgb)
-                styleMap.applyStyles(builder: builder, styles: line.styles, selColor: selArgb, highlightColors: highlightsArgb)
-                textLine = builder.build(fontCache: renderer.fontCache)
+                styleMap.applyStyles(builder: builder, styles: line.styles)
+
+                for annotationType in annotationsToBeDrawn {
```

Comment:
This for loop seems super hard to understand. Can't we chop it into smaller methods? 😄

Reply:
I agree that there's some room for improvement here. Certainly the color picking could easily be a helper function. I haven't totally wrapped my head around the actual logic yet, but something that occurs to me is that we could have some little struct like `AnnotationHelper` that has a method `annotationsForLine`, that handles a bunch of the logic and just gives us an iterator or something to work with?

Anyway it seems like the logic here is pretty straight forward, and that's really encouraging.

Reply:
Totally agree about putting this into helper functions. 
Not entirely sure about having a `annotationsForLine` helper function. I don't think it would be very performant since the function would for each line have to go through the annotations (and worst case all of them) from the beginning to find the ones shown in the current line. Oh unless the function returns an iterator for each line.

Reply:
yea, that was my thinking; maybe you init the helper with a line range, and then it tracks the state for you and for each line returns an iterator.